pkg/pub_package_reader/lib/src/package_folder.dart

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert';
+
+import 'tar_utils.dart';
+
+/// Helper methods to access package content.
+abstract class PackageFolder {
+  /// Returns and implementation that uses archive.tar.gz to provide access to
+  /// the package archive.
+  ///
+  /// Caches list of files for faster access.
+  static PackageFolder tarballPath(String path) =>
+      _ArchiveTarGzPackageFolder(path);
+
+  /// List all the file names in the directory.
+  ///
+  /// Throws IOException if directory listing or file access fails.
+  Future<List<String>> listFileNames();
+
+  /// Whether the file on the given relative [path] exists.
+  ///
+  /// Throws IOException if directory listing or file access fails.
+  Future<bool> exists(String path);
+
+  /// Read the content of the file on the given relative [path].
+  ///
+  /// Emits IOException if directory listing or file access fails.
+  Stream<List<int>> read(String path);
```

Comment:
May I suggest that we just return `Uint8List`? and if the file doesn't exist we return `null`? It's a lot simpler?

Reply:
Then none of the methods here can throw any exceptions, we just return `null` if the file doesn't exist..

Reply:
If we have I/O error, then we throw it as a fatal error and crash..